Slicing a ParquetFile does not modify initial ParquetFile.

yohplala opened the pull request:
Hello @martindurant,

I may be wrong, but given the way ``ParquetFile.__getitem__()`` is written, I think there is a bug that I correct here.
It all depends what is expected from this method:
- is it expected that it modifies current ``ParquetFile`` instance: this is what does current implementation
- or not, this is what proposes this PR by doing a deepcopy before slicing the row group list.

Given the way this method is written, by using a variable ``new_pf``, I think the 2nd way has been intended, but I may be wrong.
If the 1st way was intended, I don't think it is needed to use ``new_pf`` and return it. We could just modify ``fmd.row_groups`` in-place (meaning, no need to copy ``self`` with ``copy.copy``).

The 2nd way has been the one I expect, and I thus had some 'hard time' finding what has been going wrong in my code, with row groups mysteriously disappearing while I was using a `ParquetFile` instance, but not modifying explicitly its row groups.
([oups](https://github.com/yohplala/oups) under way :))

If I am right and this fix makes sense, a test is also enclosed.
Otherwise, if I am wrong in my guess, please ignore this PR.

PS: I had to modify `__setstate__` because `file_path`s re-appear as bytes during the `deepcopy`. But honestly, I have no great understanding what `__setstate__` does, and where does come from this `state` parameter? So please, if you accept this PR, you may want to double check this part to make sure it makes sense.

Bests

martindurant commented on fastparquet/api.py:
```diff
@@ -918,6 +918,20 @@ def __getstate__(self):
     def __setstate__(self, state):
         self.__dict__.update(state)
+        # Decode 'file_path'.
+        fmd = self.fmd
```
in this case, you are not editing a list of objects, but attributes of an object deep in the list, so you can work on `self.fmd` directly without this line and `self.fmd = ` later on.

martindurant commented:
Yes, I believe you are completely correct in your description and the the code here. I posted one comment to simplify the new block a little.

I believe deepcopy must call getstate (on the original) and setstate (on the new copy), which includes the fmd ThriftObject, which in turn defines `__deepcopy__` as `pickle.loads(pickle.dumps(self))` - this is perhaps an unnecessarily expensive way to duplicate all the dictionaries contained. Cycling via the binary representation is what causes the paths to reset to bytes; if `__deepcopy__` had a better implementation, the new code in setstate wouldn't be needed.

The following seems like it should work (untested):
```
--- a/fastparquet/cencoding.pyx
+++ b/fastparquet/cencoding.pyx
@@ -772,8 +772,9 @@ cdef class ThriftObject:
         return self.copy()

     def __deepcopy__(self, memodict={}):
-        import pickle
-        return pickle.loads(pickle.dumps(self))
+        import copy
+        d = copy.deepcopy(self.data)
+        return ThriftObject(self.name, d)

     cpdef _asdict(self):
```

yohplala commented:
> The following seems like it should work (untested):
> 
> ```diff
> --- a/fastparquet/cencoding.pyx
> +++ b/fastparquet/cencoding.pyx
> @@ -772,8 +772,9 @@ cdef class ThriftObject:
>          return self.copy()
> 
>      def __deepcopy__(self, memodict={}):
> -        import pickle
> -        return pickle.loads(pickle.dumps(self))
> +        import copy
> +        d = copy.deepcopy(self.data)
> +        return ThriftObject(self.name, d)
> 
>      cpdef _asdict(self):
> ```

Thanks for the feedback!
I am sorry I don't understand: should I try your fix?
(I can, no worries, just to make sure I am correctly understanding)
Thanks for your help again!

martindurant commented:
> I am sorry I don't understand: should I try your fix?

Yes, I would appreciate it, I think it's a better fix in performance and also aesthetically better.

yohplala commented:
> Yes, I would appreciate it, I think it's a better fix in performance and also aesthetically better.

ok, done.
Feeling a bit 'challenged' on speed (kidding here ;)), I also reworked the 'de-' binarization of `file_path`.
I moved the 3 `if` within the `for` loop so that it is just checked on the 1st row group once before the `for` loop.
I think it makes sense because this `file_path` is here when we have `file_scheme='hive'`, ie.e it will then apply on all row groups.

Do you think this is acceptable?

If yes, I could also port the fix to `ParquetFile._parse_header()` as this piece of code comes from here actually.

yohplala commented:
Woh, that is weird, all ubuntu are failing, windows are ok.
I am on ubuntu and it is working on my machine.
Deep diving.

yohplala commented:
Ah, I think maybe you can advise me: what does "wheel" mean? Are there some kind of pre-built binaries somewhere that need to be updated?

martindurant commented:
Wheels are the binary distribution of packages used by pip. In the wheel CI runs, wheels are built using "cibuildwheel" and then installed for testing, rather than installing (pip install -e .) and testing directly from the repo directory.

yohplala commented:
> Wheels are the binary distribution of packages used by pip. In the wheel CI runs, wheels are built using "cibuildwheel" and then installed for testing, rather than installing (pip install -e .) and testing directly from the repo directory.

@martindurant I reverted the change you proposed in `cencoding.ThriftObject.__deepcopy_` in [this other branch](https://github.com/yohplala/fastparquet/tree/rgs_deepcopy_dummy), and wheels are now ok.
Do you have an advise?

Thanks in advance for your support! (I am lost! :))

martindurant commented:
```
>               chunk[1] = chunk.get(1).decode()
E               AttributeError: 'str' object has no attribute 'decode'
```
but it's no longer a bytes instance. Tricky, because it *is* a bytes when you get to setstate via unpickle, but not if you get there via my version of deepcopy. No idea why this only shows up in some test runs.

I'm not immediately what to do about it. We could stick with your version and iterate maybe.

yohplala commented:
> We could stick with your version and iterate maybe.

I 'pushed' the 'single-if-concept' to its max :)
It's certainly a bit hacky, but it is done on a single row group.
Let's see if it works.
